test/integration/emulator/FakeVideoDecoder.h

```diff
@@ -0,0 +1,244 @@
+
+            _packets.push_back(*data);
+        }
+
+        decodeAssembledFrames(timestamp);
+    }
+
+    void resetPacketCache()
+    {
+        FAKE_VIDEO_DECODER_LOG("Reset packet cache (size = %zu)", _loggableId.c_str(), _packets.size());
+        _packets.clear();
+    }
+
+    Stats getStats() const { return _stats; }
+
+private:
+    void decodeAssembledFrames(const uint64_t timestamp)
+    {
+        while (_assembledFrames.size())
+        {
+            auto frameNum = _lastDecodedFrameNum + 1;
+            auto nextFrame = std::find_if(_assembledFrames.begin(),
+                _assembledFrames.end(),
+                [frameNum](const AssembledFrame& frame) { return frame.frameNum == frameNum || frame.keyFrame; });
+
+            if (nextFrame != _assembledFrames.end())
+            {
+                FAKE_VIDEO_DECODER_LOG("decoded frame: %d, from %zu",
+                    _loggableId.c_str(),
+                    nextFrame->frameNum,
+                    nextFrame->endpointIdHash);
+
+                _lastDecodedFrameNum = nextFrame->frameNum;
+                _ssrc = nextFrame->ssrc;
+
+                auto ssrc = _ssrc;
+                auto reorderQueueSize = std::count_if(_assembledFrames.begin(),
+                    _assembledFrames.end(),
+                    [frameNum, ssrc](
+                        const AssembledFrame& frame) { return (frame.frameNum > frameNum) && (ssrc == frame.ssrc); });
+
+                updateStats(nextFrame->endpointIdHash, nextFrame->tag, reorderQueueSize, timestamp);
+
+                _assembledFrames.remove_if([nextFrame](const AssembledFrame& frame) {
+                    return frame.ssrc != nextFrame->ssrc || frame.frameNum <= nextFrame->frameNum;
+                });
+            }
+            else
+            {
+                break;
+            }
+        }
+    }
+
+    void removeDecodedPackets(uint32_t frameNum)
+    {
+        auto elementsToErase = std::remove_if(_packets.begin(),
+            _packets.end(),
+            [frameNum](const fakenet::FakeVideoFrameData& data) { return data.frameNum == frameNum; });
+        if (elementsToErase != _packets.end())
+        {
+            _packets.erase(elementsToErase);
+        }
+    }
+
+    bool assembleFrame(const fakenet::FakeVideoFrameData& data)
+    {
+        std::list<fakenet::FakeVideoFrameData> framePackets;
+        framePackets.push_back(data);
+
+        std::copy_if(_packets.begin(),
+            _packets.end(),
+            std::back_inserter(framePackets),
+            [&data](const fakenet::FakeVideoFrameData& packet) {
+                return packet.frameNum == data.frameNum && packet.ssrc == data.ssrc;
+            });
+
+        framePackets.sort([](const fakenet::FakeVideoFrameData& one, const fakenet::FakeVideoFrameData& another) {
+            return one.packetId < another.packetId;
+        });
+
+        const auto numPackets = framePackets.size();
+        const bool result =
```

Comment:
This means something other than "result". It is true when the frame is complete and contains all packets in the frame, right?

Reply:
Agree.